### ISSUE TITLE
Convert to fake image badly but reconstruct well 

### ISSUE THREAD

**Issue by poppopting** [junyanz/pytorch-CycleGAN-and-pix2pix#1057]
Hi, 
Currently i'm trying to implement CycleGAN on summer2winter dataset,  and model structure and hyper parameters  like learning rate and lambda are all the same as paper described (9 residual blocks). However, after training for 200 epochs, i found that when using generators to convert to fake images, the results are weird . But if i reconstructed them by the other generator, the results are close to original images. 
I'm wondering whether the problem is in the discriminator since the results may shows the problem for discriminator in distinguishing real and fake images.  
In order to improve model, should i  try to make discriminator more powerful ? or  to add identity loss for training ? (for summer2winter dataset, it seems authors didn't use identity loss to train the model, how it would be if i add identity loss for training ? )
[training result](https://ibb.co/921wHtx)

**Comment by junyanz** [comment]
Does the pre-trained model work for you? (see the model downloading [script](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/scripts/download_cyclegan_model.sh))  I think that we probably didn't use the identity loss. @taesungp

**Comment by taesungp** [comment]
hmm are they visually considerably worse when you compare it with the results on the project webpage? you can try training with identity loss too.

**Comment by poppopting** [comment]
per-trained model works for me ! But I'm still want to implement it in my way and adjust hyper-parameter to the differences . i found the structure in my script is a little different from yours yesterday, so i adjust them and retrain again, maybe the results would good after training !  thanks for answering :)) !!!

**Comment by junyanz** [comment]
If you are implementing your own cyclegan, I recommend that you can do it step by step. For example, you can use part of our code base first (e.g., generator and discriminator),  and test the part you wrote (loss, training).

**Comment by poppopting** [comment]
Okay, i got it.  I appreciate your help very much !!!

**Comment by superkido511** [comment]
@poppopting Did you mangage to solve the issue? I'm facing the same issue now. I'm trying to use cycle gan for manga colorization (without coverting color spaces). Generated images are very bad, while reconstructed images are very good. I used the original code base. Here is my training opts:
 python3 train.py --dataroot ./datasets/unprocessed --name manga_cyclegan4 --model cycle_gan --no_dropout --loadSize 512 --fineSize 256 --lr_decay_iters 100 --niter_decay 200 --save_epoch_freq 30 --display_freq 1000 --which_model_netD n_layers --display_id 0 --resize_or_crop scale_width_and_crop --lambda_identity 0.1
Can you help me please prof @junyanz ?

**Comment by poppopting** [comment]
@superkido511  yes, I have solved the problems. But in my case, there are some typos in the Discriminator written by me and the problem was solved after I corrected them. 
But in your case, you used pre-trained model, maybe the problems are just in your parameter setting. Does the "bad" in your generated images is they are hardly colored? or the images are painted in the wrong color?

**Comment by superkido511** [comment]
@poppopting Hello, I didn't use pretrained model. By "bad", I mean some part of the image are colored, but most of the time, it's the wrong color. I also noticed that after about 100, the discriminator loss of domain A(classify uncolored images) are nearly 0 (0.001 ~ 0.002) most of the time and occasionally go up (0.1 ~ 0.2). Here are my training result:
![image](https://user-images.githubusercontent.com/26477048/92068412-079e8b00-edd1-11ea-9b77-3b6e9fe6c7b1.png)
You can see that the reconstructed images look very close to the real images, while the generated images are quite bad. I tried to lower cycle loss weight, make generator more powerful, use separate learning rates, optimizers for each networks. Nothing seems to work so far

**Comment by poppopting** [comment]
In my case, I made a mistake in the below part and got both weak generator :
```(correct version)
loss_D_real = self.criterionGAN(pred_real, True)
loss_D_fake = self.criterionGAN(pred_fake, False)
```
But I think you only have a problem with generator A (it seems generator B is good),  your situation may differ from mine.
Besides, I found there is a repo fitting CycleGan with manga colorization,  [link](https://github.com/OValery16/Manga-colorization---cycle-gan), hope it can help you.

**Comment by superkido511** [comment]
@poppopting Thank you so much !. I'll try it and see how it goes

**Comment by qtjv** [comment]
> @superkido511 In my case, I made a mistake in the below part and got both weak generator :
> 
> ```
> #correct version
> loss_D_real = self.criterionGAN(pred_real, True)
> loss_D_fake = self.criterionGAN(pred_fake, False)
> ```
> 
> But I think you only have a problem with generator A (it seems generator B is good), your situation may differ from mine. Besides, I found there is a repo fitting CycleGan with manga colorization dataset: [link](https://github.com/OValery16/Manga-colorization---cycle-gan), hope it can help you.

I found no errors in these two sentences in the author's code

**Comment by poppopting** [comment]
> > @superkido511 In my case, I made a mistake in the below part and got both weak generator :
> > ```
> > #correct version
> > loss_D_real = self.criterionGAN(pred_real, True)
> > loss_D_fake = self.criterionGAN(pred_fake, False)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > But I think you only have a problem with generator A (it seems generator B is good), your situation may differ from mine. Besides, I found there is a repo fitting CycleGan with manga colorization dataset: [link](https://github.com/OValery16/Manga-colorization---cycle-gan), hope it can help you.
> 
> I found no errors in these two sentences in the author's code

Yes, there is no errors in the author's code, the code I put previously was the error made by me when I want to build this CycleGAN model from scratch

**Comment by lihanqing03021** [comment]
> [@poppopting](https://github.com/poppopting) Hello, I didn't use pretrained model. By "bad", I mean some part of the image are colored, but most of the time, it's the wrong color. I also noticed that after about 100, the discriminator loss of domain A(classify uncolored images) are nearly 0 (0.001 ~ 0.002) most of the time and occasionally go up (0.1 ~ 0.2). Here are my training result: ![image](https://user-images.githubusercontent.com/26477048/92068412-079e8b00-edd1-11ea-9b77-3b6e9fe6c7b1.png) You can see that the reconstructed images look very close to the real images, while the generated images are quite bad. I tried to lower cycle loss weight, make generator more powerful, use separate learning rates, optimizers for each networks. Nothing seems to work so far

do you solve the issue? i have the same issue
